service/topology_coordinator.cc

```diff
@@ -850,6 +853,90 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+
+            while (true) {
+                topology& topo = _topo_sm._topology;
+                const utils::UUID& global_request_id = topo.global_request_id.value();
+                const table_id& table_id = topo.truncate_table_id.value();
+                const std::chrono::milliseconds timeout = topo.truncate_table_timeout_ms.value();
+                lw_shared_ptr<replica::table> table = _db.get_tables_metadata().get_table_if_exists(table_id);
+
+                std::optional<sstring> error;
+                if (table) {
+                    const sstring& ks_name = table->schema()->ks_name();
+                    const sstring& cf_name = table->schema()->cf_name();
+
+                    rtlogger.info("Performing TRUNCATE TABLE global topology request for {}.{} with timeout {}", ks_name, cf_name, timeout);
+
+                    const locator::tablet_map& tmap = get_token_metadata_ptr()->tablets().get_tablet_map(table_id);
+
+                    // Collect the IPs of the hosts with replicas
+                    std::unordered_set<gms::inet_address> replica_endpoints;
+                    co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& tinfo) {
+                        for (const locator::tablet_replica& replica: tinfo.replicas) {
+                            const std::set<gms::inet_address> addrs = _gossiper.get_nodes_with_host_id(replica.host);
+                            replica_endpoints.insert(addrs.begin(), addrs.end());
+                        }
+                        return make_ready_future<>();
+                    });
+
+                    // Check if all the nodes with replicas are alive
+                    const std::set<gms::inet_address> live_nodes = _gossiper.get_live_members();
+                    const size_t live_replicas = std::count_if(replica_endpoints.begin(), replica_endpoints.end(),
+                                                                [&live_nodes] (const gms::inet_address& addr) {
+                                                                    return live_nodes.contains(addr);
+                                                                });
+                    if (live_replicas == replica_endpoints.size()) {
+                        try {
+                            // Send the RPC to all replicas
+                            const netw::messaging_service::clock_type::time_point timeout_deadline = lowres_clock::now() + timeout;
+                            const service::frozen_topology_guard frozen_guard { _topo_sm._topology.session };
+                            co_await coroutine::parallel_for_each(replica_endpoints, [&] (const gms::inet_address& ep) -> future<> {
+                                netw::messaging_service::msg_addr addr(ep);
+                                co_await ser::storage_proxy_rpc_verbs::send_truncate(&_messaging, addr, timeout_deadline, ks_name, cf_name, frozen_guard);
+                            });
```

Comment:
I don't think we should do RPCs while holding the guard, it may block group0 for substantial amount of time.

It would also prevent us from canceling the request via group0 should we want to add such ability.